newapi/ooniapi/probe_services.py

```diff
@@ -41,6 +43,27 @@ def generate_report_id(test_name, cc: str, asn_i: int) -> str:
     return rid
 
 
+def get_probe_ip() -> str:
+    real_ip_headers = [
+        "X-Forwarded-For",
+        "X-Real-IP"
+    ]
+    for h in real_ip_headers:
+        if h in request.headers:
+            return request.headers.getlist(h)[0].rpartition(' ')[-1]
+
+    return request.remote_addr
+
+def lookup_probe_asn(ip: str) -> str:
+    with geoip2.database.Reader(current_app.config["GEOIP_ASN_DB"]) as reader:
```

Comment:
Do we know how long it takes to open and close the databases every time? Perhaps they should be kept open instead.

Reply:
Yeah according to the docs of geoip2 it's recommended to keep a long lived handle on the open database as the process of loading it via the reader is quite IO heavy.

I was originally planning to store these handles in the `g` global flask state, but then realised that it's [only preserved across requests](https://flask.palletsprojects.com/en/2.0.x/api/#flask.g), so it's not really going to help much. Do you have any ideas on how to store some long lived reference to these handles that is local to each flask thread/process?

Reply:
Ok nvm, I figured this out. It seems like I just have to attach to the app context which is shared across requests, but thread local.